secure/json-rpc/src/lib.rs

```diff
@@ -0,0 +1,261 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+//! The purpose of the LightweightJsonClient presented here is to provide a lightweight and secure
+//! JSON RPC client to talk to the JSON RPC service offered by Libra Full Nodes. This is useful
+//! for various security-critical components (e.g., the secure key manager), as it allows
+//! interaction with the Libra blockchain in a minimal and secure manner.
+//!
+//! Note: While a JSON RPC client implementation already exists in the Libra codebase, this
+//! provides a simpler and (hopefully) more secure implementation with fewer dependencies.
+#![forbid(unsafe_code)]
+
+use crate::error::Error;
+use libra_types::{
+    account_address::AccountAddress, account_state::AccountState,
+    account_state_blob::AccountStateBlob, transaction::SignedTransaction,
+};
+use serde::{Deserialize, Serialize};
+use serde_json::{json, Value};
+use std::convert::TryFrom;
+use ureq::Response;
+
+mod error;
+#[cfg(test)]
+mod tests;
+
+/// Various constants for the JSON RPC client implementation
+const JSON_RPC_VERSION: &str = "2.0";
+const REQUEST_TIMEOUT: u64 = 10_000;
+
+/// This defines a simple and generic interface by which to interact with the JSON RPC service
+/// offered by each Libra Full Node.
+pub trait JsonRpcClient {
+    /// Submits a signed transaction to the Libra blockchain via the JSON RPC API.
+    fn submit_signed_transaction(&self, signed_transaction: SignedTransaction)
+        -> Result<(), Error>;
+
+    /// Returns the associated AccountState for a specific account at a given version height
+    /// using the JSON RCP API.
+    fn get_account_state(
+        &self,
+        account: AccountAddress,
+        version: u64,
+    ) -> Result<AccountState, Error>;
+}
+
+/// Provides a lightweight JsonRpcClient implementation.
+struct LightweightJsonClient {
+    host: String,
+}
+
+impl LightweightJsonClient {
+    // TODO(joshlind): remove this flag once the client is integrated into key manager. Otherwise,
+    // rust complains new() is never used.
+    #[cfg(test)]
+    pub fn new(host: &str, port: u16) -> Self {
+        let host = format!("http://{}:{}", host, port);
+        Self { host }
+    }
+
+    // Executes the specified request method using the given parameters by contacting the JSON RPC
+    // server.
+    fn execute_request(&self, method: String, params: Vec<Value>) -> Response {
+        let request_id = 0;
+        ureq::post(&self.host)
+            .timeout_connect(REQUEST_TIMEOUT)
+            .send_json(
+                json!({"jsonrpc": JSON_RPC_VERSION, "method": method, "params": params, "id": request_id}),
+            )
+    }
+
+    // Sends a submit() request to the JSON RPC server using the given transaction.
+    fn submit(&self, signed_transaction: SignedTransaction) -> Result<Response, Error> {
```

Comment:
submit_transaction?

Reply:
I was explicitly using the request method names to label these functions to make it obvious (e.g., submit and get_account_state_with_proof), however, I've changed this to submit_transaction as you suggest. I don't have super strong feelings.